Fix issues with resuming async tasks awaiting a future (backport #1469)

bjsowa opened the pull request:
Backport of #1469 to Jazzy

bjsowa committed:
Fix issues with resuming async tasks awaiting a future (backport #1469)

Signed-off-by: Błażej Sowa <bsowa123@gmail.com>
Signed-off-by: Nadav Elkabets <elnadav12@gmail.com>
Co-authored-by: Nadav Elkabets <32939935+nadavelkabets@users.noreply.github.com>

bjsowa committed:
Reorder task execution in Executor to maintain LIFO for backward compatibility

Signed-off-by: Błażej Sowa <bsowa123@gmail.com>

bjsowa commented:
@nadavelkabets I modified the deque logic to preserve the LIFO order which is maintained for backward compatibility in Jazzy. Could you check it?

nadavelkabets commented on rclpy/rclpy/executors.py:
```diff
@@ -581,21 +606,32 @@ def _wait_for_ready_callbacks(
+            for _ in range(ready_tasks_count):
+                # We pop from the right to maintain LIFO order for backward compatibility
+                # From Kilted onwards, the order of task execution is FIFO
+                task = self._ready_tasks.pop()
```
I believe it's more complicated to mimic the old behavior precisely. 
While iterating over ready tasks, new tasks may get added to the ready tasks list by either a user that created a new task or a finished task that triggered done callbacks.
In the existing implementation, these would only execute in the next _wait_for_ready_callbacks iteration (they iterate over a copy of the tasks list) while in our implementation they would execute right away (we iterate over the same list).
In addition, prior to our changes blocked tasks stayed in the tasks list while we remove and append them  back. This also means that we change the execution order of resumed tasks.

The only perfect solution I can imagine is to add a creation_time timestamp to the TaskData class, and using a heapq for the ready_tasks list to keep it sorted from oldest task to newest task. Furthermore, we currently utilize done callbacks to execute resumed tasks. To maintain the same order, we would have to give the done callback the same timestamp as the original task.
I honestly find the idea of switching Jazzy to FIFO order more realistic and straightforward.
Is that unreasonable?

bjsowa replied:
If changing the order to FIFO is an option, I'm all for it.

nadavelkabets replied:
I couldn't attend the working group meeting this week but it appears like they're fine with the switch to FIFO ordering.

> Fix issues with resuming async tasks awaiting a future (backport #1469)
https://github.com/ros2/rclpy/pull/1560
Try a full CI, to check if we are breaking something by modifying this internal behavior.
In other repos we went ahead with similar changes since people shouldn’t rely on this internal ordering of execution for wait-set based executors.

Flova committed:
Increase clock accuracy (#1564)

Signed-off-by: Błażej Sowa <bsowa123@gmail.com>

kscottz commented:
We discussed this in our weekly project meeting. We've added your issue to the next [Client Library Working Group Meeting on January 9th](https://docs.ros.org/en/jazzy/The-ROS2-Project/Governance.html) for discussion and [you have been added to the agenda](https://docs.google.com/document/d/1-tU2AKoHKLCVHPBz0QchYownmZZvYjHmVUQXT2WBMOw/edit?usp=sharing).

bjsowa committed:
Revert "Reorder task execution in Executor to maintain LIFO for backward compatibility"

This reverts commit c388647edf3cc4357d7de3ec2d6c511a670c8de8.

Signed-off-by: Błażej Sowa <bsowa123@gmail.com>

bjsowa committed:
Fix failing test

Signed-off-by: Błażej Sowa <bsowa123@gmail.com>

bjsowa commented:
I reverted the changes related to maintaining LIFO order and modified the test that relied on ordering. I think it's ready to run the full CI @mjcarroll @ahcorde @nadavelkabets

nadavelkabets commented on rclpy/rclpy/executors.py:
```diff
@@ -150,6 +154,12 @@ def timeout(self, timeout):
+@dataclass
+class TaskData:
+    source_node: 'Optional[Node]' = None
+    source_entity: 'Optional[Any]' = None
```
Let's use the existing typing for entity
```suggestion
    source_entity: 'Optional[WaitableEntityType]' = None
```

nadavelkabets commented on rclpy/test/test_executor.py:
```diff
@@ -299,7 +333,39 @@ async def coroutine():
+
+                coroutine_future = executor.create_task(coroutine)
+
+                start_time = time.monotonic()
```
If I remember correctly time.monotonic caused this test to fail in windows environment and this was fixed in rolling by switching to perf_counter, take a look at #1564

bjsowa replied:
I backported the change

nadavelkabets reviewed (changes requested):
Apart from the 2 minor suggestions, LGTM.
Are tests failing for you locally? The failing tests do not seem related...

bjsowa committed:
Update rclpy/rclpy/executors.py

Co-authored-by: Nadav Elkabets <32939935+nadavelkabets@users.noreply.github.com>
Signed-off-by: Błażej Sowa <bsowa123@gmail.com>

bjsowa commented:
> Are tests failing for you locally? The failing tests do not seem related...

Nope, the tests are passing for me locally.

JWhitleyWork commented:
Looks like this also needs a backport to `humble` based on [this comment](https://github.com/ros2/rclpy/pull/1555#issuecomment-3667599861).

Thanks again for working on this!

nadavelkabets commented:
@fujitatomoya @skyegalaxy @ahcorde @mjcarroll 
Please run full CI on this

ahcorde commented:
Pulls: ros2/rclpy#1560
Gist: https://gist.githubusercontent.com/ahcorde/5de7a805edbbfe64f19e112a28b1c866/raw/92ec171f8acddb3c7c698fc860099257102cc621/ros2.repos
BUILD args:  --packages-above-and-dependencies rclpy
TEST args:  --packages-above rclpy
ROS Distro: jazzy
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/17928
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=27056)](http://ci.ros2.org/job/ci_linux/27056/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=20430)](http://ci.ros2.org/job/ci_linux-aarch64/20430/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=6824)](http://ci.ros2.org/job/ci_linux-rhel/6824/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=26388)](http://ci.ros2.org/job/ci_windows/26388/)

bjsowa commented:
@ahcorde Linux and Linux-rhel CI did not start properly due to some internal buildfarm error

JWhitleyWork commented:
Would one of you mind creating the backport for this to `humble`?

bjsowa commented:
> Would one of you mind creating the backport for this to `humble`?

I started backporting it to humble but had some issues with some tests randomly failing. I'll try to create a PR next week.

JWhitleyWork commented:
@bjsowa Thanks for the update!
